src/cargo/util/toml/mod.rs

```diff
@@ -1553,6 +1556,23 @@ impl TomlManifest {
         Ok(patch)
     }
 
+    fn patch(&self, cx: &mut Context<'_, '_>) -> CargoResult<HashMap<Url, Vec<Dependency>>> {
+        let _ = cx.config.maybe_init_patch(|| {
```

Comment:
To me at least this feels somewhat backwards in terms of how initialization/parsing of these sorts of config tables happens. This is just sort of randomly buried in unrelated code and has a very "bolted-on" feeling.

Could this be refactored so the `Config` field could be parsed on-demand instead of sort of off-on-the-side like this?

Reply:
I agree, it does feel a little bolted-on. The reason it is this way is because of `DetailedTomlDependency::to_dependency`, which takes a `Context`, which is only ever constructed "down here". I suppose we could construct a `Context` later on-demand, though that would mean the on-demand path _also_ needs to:

Populate the platform

https://github.com/rust-lang/cargo/blob/ce0b8e8574796a951ebe7ed872f1e2f8cc538901/src/cargo/util/toml/mod.rs#L1227-L1228

Deal with `to_dependency` changing `Context::nested_paths`:

https://github.com/rust-lang/cargo/blob/ce0b8e8574796a951ebe7ed872f1e2f8cc538901/src/cargo/util/toml/mod.rs#L1785

And extract warnings and errors and push them as appropriate:

https://github.com/rust-lang/cargo/blob/ce0b8e8574796a951ebe7ed872f1e2f8cc538901/src/cargo/util/toml/mod.rs#L1371-L1376

That's not insurmountable, but it would be some code repetition. It would also mean that warnings/errors in `.cargo/config`'s `[patch]` would be handled at a different "time" in the process than those in `Cargo.toml`, which may in turn mean that a patch in one or the other manifest very differently. Not sure how much of a problem that is?

The other way to look at this is that this is the _right_ place for initialization to happen: all patch sections, regardless of where they come from, get processed at the same time. We just need to store the config patches somewhere separate so that we can get them back later. It only really looks weird because of the fact that we store it in `Config`. If we instead store the config patch in the `Manifest` (alongside but not merged with) the regular `[patch]`, it'd be less of a problem. The downside with _that_ is that we'd then have to modify more code paths to ensure that they carry along this extra manifest information all the way to where we apply the patches. But maybe that's a nicer approach?

Reply:
Yeah I understand the wonkiness today, and in some sense this wonkiness is beacuse this feature is going against how Cargo is designed (and hence how the code is structured). I don't really have a preference on how to organize this but I would basically just prefer that refactoring was done (if necessary), to give this less of a bolted on feeling.

Reply:
I have a different approach a go in https://github.com/rust-lang/cargo/pull/9204/commits/6995e1dd1b89d919aa0e2f8443c7cbaf4931bb32 that I think you'll like better. Let me know what you think!